types/dispatchr/index.d.ts

```diff
@@ -0,0 +1,66 @@
+// Type definitions for dispatchr 1.2
+// Project: https://github.com/yahoo/fluxible#readme
+// Definitions by: Ragg <https://github.com/Ragg->
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
+/// <reference types="node" />
+import { EventEmitter } from 'events';
+
+export interface DispatcherState {
+    stores: { [storeName: string]: any };
+}
+
+export interface DispatcherOption {
+    stores?: StoreClass[];
+    errorHandler?: (e: DispatcherError, context: DispatcherContext) => void;
+}
+
+export interface StoreClass {
+    storeName: string;
+    new(): Store;
+}
+
+export interface Store<S = {}> extends EventEmitter {
+    dehydrate?(): S;
+    rehydrate?(state: S): void;
+    shouldDehydrate?(): boolean;
+    emitChange(): void;
+}
+
+export interface Dispatcher {
+    createContext(context: {}): DispatcherContext;
+    registerStore(store: StoreClass): void;
+    isRegistered(store: StoreClass | string): boolean;
+    getStoreName(store: StoreClass | string): string;
+}
+
+export interface DispatcherInterface {
+    getContext(): DispatcherContext;
+    getStore: DispatcherContext['getStore'];
+    waitFor: DispatcherContext['waitFor'];
+}
+
+export interface DispatcherContext {
+    getStore(name: string): Store;
+    getStore<T extends StoreClass>(name: T): T;
+
+    dispatch(actionName: string, payload: any): void;
```

Comment:
I wonder if there's a nice way to strongly type this...

Random idea that probably only works in 2.8 *if* it works at all:

```ts
interface DispatcherContext<ActionTypeMap extends object|string> {
  dispatch <Name extends (ActionTypeMap extends string ? string : keyof ActionTypeMap)>(
    actionName: Name, payload: ActionTypeMap extends string ? any : ActionTypeMap[Name]
  ): void
}
```

```ts
interface TypeMap {
  ACTION_FOO: { foo: string }
  ACTION_BAR: { bar: string }
}
declare var ctx: DispatcherContext<TypeMap>
ctx.dispatch('ACTION_FOO', { foo: 'foo' })
ctx.dispatch('ACTION_BAR', { foo: 'foo' }) // type error, expected { bar: string }
ctx.dispatch('CTHULHU', { foo: 'foo' }) // type error, expected never
```

Reply:
Great :+1:
But since TypeScript 2.8 has not been officially released yet, I'd like to avoid putting it in at the moment😔